### PR TITLE
Added FindCudaArch

### PR DESCRIPTION
**Pull request by borisfom**
I have added FindCudaArch.cmake - there is currenltly 2+ copies of this file as well, I believe it belongs here. After the use of this repo will go into torch distro, we could remove copies from individual modules.

### CODE REVIEW

**Comment by hughperkins** [comment]
Sorry, I dont have default 'watch' on my repos, and didnt see this before...

**Comment by hughperkins** [comment]
What does this do? How can I test it is doing what it is intended to do?

**Comment by borisfom** [comment]
This PR moves FindCudaArch.cmake from cmake/directory of cutorch. Reason: it was being duplicated in cunn, etc. Now those modules won't even have their own cmake/ dirs.

To check: integrate this PR and one from cutorch: https://github.com/torch/cutorch/pull/334 into your workspace and see if it breaks. Oh you will also need https://github.com/torch/torch7/pull/55.

**Comment by hughperkins** [comment]
Do you know you closed this PR?  Is that intentional?

**Comment by hughperkins** [comment]
Sorry, I mean, what does FindCudaArch do?  How can I test that it is/isnt present and working ok?

**Comment by borisfom** [comment]
Hi! I don't know who closed it :)
FindCudaArch tries to determine current CUDA 'compute' arch of the host (like 3.0 vs 5.0 vs 5.2) and optimize NVCC flags to only compile for that arch; or, if it's explicuitly overridden by user to 'All', then use that. So I believe it belongs to FindCUDA and should ultimately go to CMake. Will talk to James about it.

To make a real thougough check, one would :
1. pre-patch: build Torch with VERBOSE=1, save log to a file (make sure you see NVCC command-lines there)
2. apply this PR only, rebuild, make sure command lines are the same
3. apply the other 2 PRs above, rebuild, check command lines.

**Comment by hughperkins** [comment]
I'd prefer something more like a unit-test to be honest.  Doesnt have to be automated, but you can see for example that for the original FindCUDA I did quite a lengthy test procedure: https://github.com/torch/cutorch/pull/256#issuecomment-179815577  The tests dont have to be quite so lengthy, but maybe eg a CMakeLists that prints the architecture or something?

**Comment by hughperkins** [comment]
Sent you a PR with some kind of draft idea of how to validate this.  Thoughts?

**Comment by borisfom** [comment]
I can't think of proper validation actually: having it print out resulting architectures for the user reality check should be enough.
Also, I have contacted James with suggestion to merge this into FindCUDA - let's see what he says.

**Comment by borisfom** [comment]
OK this should have been a separate PR: 
I generalized the module to not depend on Torch and being installable via cmake.
This was particularly useful for Android port (as Android build relies on native Torch be on the system and does not use rocks).
Also, it would widen the application of this module a lot. Some systems have cmake that is not CUDA aware at all, btw.
README also updated. Let me know what you think!

**Comment by hughperkins** [comment]
Ok, sounds good :-)  Do you want to merge in my PR into your branch? :-)  This is how I validated that FindCudaArch seems to do more or less what it should be doing, and I think it'd be good to keep such validation in the repo.

File: README.md [hunk 1]
```diff
@@ -28,23 +33,20 @@ software)
 
 ### Why not just use the Ubuntu standard FindCUDA?
 
-- doesnt contain the fix for incremental builds on cutorch
-
-### Why not just download the gz directly?
-
-- doesnt contain the fix for incremental builds on cutorch
+- Ubintu comes with CMake 2.8 which has the dependency issue with CUDA .cu files.
```
Comment (hughperkins) [inline README.md:36]: typo

File: README.md [hunk 2]
```diff
@@ -28,23 +33,20 @@ software)
...
+Expecting official fix to come out with CMake 3.51 - not sure about 2.8 backport.
+This repo might evolve appropriately as a function of this, or disappear eventually, as
 and when upstream is fixed, and a new version of cmake is available
 
 ## License
 
 MIT
 
-## Where can I find out more about FindCUDA?
+## Where can I find out general informtion about FindCUDA module of CMake?
```
Comment (hughperkins) [inline README.md:49]: typo

Comment (borisfom) [inline README.md:49]: Please feel free to alter my language as you see fit :)

**Comment by borisfom** [comment]
BTW, there is some interesting development/news from James: 

Make has a mechanism for downloading stuff as part of configure. I'll have to look up exactly how to do this. 

James

On Feb 18, 2016, at 7:00 PM, Boris Fomitchev bfomitchev@nvidia.com wrote:

Hi James,

Thanks for the clarification!
As I mentioned in the other thread,  requiring cmake  upgrade may not be the option for Torch. 
The other option of refactoring our little FindCUDA module to fetch the latest code from cmake repo instead of keeping its own, actually seems more viable.

Re FindCudaArch:  right, Torch didn’t ask us to fix this. They used the same mechanism as FindCUDA though, to shadow, if you will, nonexistent cmake feature instead of a bug.
I just can’t help noticing there is a substantial piece of generic Cmake functionality directly related to Nvidia tools only,  
that is being duplicated many times even with the same outside project (Torch) and across projects (it actually was taken from Caffe). 
This code is bound to get forked and outdated, not knowing latest compute architectures (I actually added 5.0 and 5.2 to it myself)  etc unless it’s adopted into common Cmake module. 
Isn’t that exactly how new modules/features/extensions get to Cmake normally ?

Since it is very closely related to FindCUDA which is already multiple files module, I think it can be simply merged into it. Here:
https://github.com/borisfom/FindCUDA/blob/master/FindCudaArch.cmake

Best,
-Boris.

File: README.md (hunk 2 above)
Comment (hughperkins) [inline README.md:49]: well, `informtion` => `information`

File: README.md (hunk 1 above)
Comment (hughperkins) [inline README.md:36]: 'Ubintu' => 'Ubuntu'

**Comment by borisfom** [comment]
Merged changes from James' repo.
Added Android compilation
Added configurations for NVidia Shield tablets

**Comment by hughperkins** [comment]
Ok.  Please can you create some simple test, that I can run, so I can see it's all working :-)

Also, the more stuff you add into one PR, the more I look at it, and think, 'ummm.... this looks like a lot of stuff to check :-)'

Alternatively, cos I know I've already claimed `FindCUDA` in the rocks file, but you could create eg `FindCudaAll`, with all the other bits in.  I think you know more about them than me, and might be better placed to maintain them?

(PS I really dislike building cutorch, so if you can make the test(s) build some small tiny dummy project, that will be much nicer for me :-)  Nicer == quicker merge :-D)

**Comment by borisfom** [comment]
Sorry Hugh, I am up to my head with the Torch work, no tests...  This is direct copy of PRs I am sending to the primary CMake repo here: https://github.com/Kitware/CMake/pulls

So one way to handle this here would be to wait until they merge my changes, and then replace actual .cmake files with FILE() references to the top of their tree - this way they will always be tested and up-to-date. 
In fact, you can already do it now- I have verified that their fix for dependencies work as well. So you will have the same functonality  now, and you'll get my changes when they merge.

**Comment by hughperkins** [comment]
Yeah, me too.  So, the situation is, currently FindCUDA works, as-is, does what it says on the can.  If I merge, it might stop working, and then I have to do lots of digging, and try to fix it.  Or, I might merge it, and it broke it, but I dont notice until people start complaining it's broken, which isnt so bad, but invariably happens right when I have something really urgent and time-critical to do at the same time :-P

But at the same time you're adding lots of stuff, and it looks really useful.

Hence, why I'm thinking, maybe you create `FindCudaAll`? Sure, the end result might be that everyone uses your repo, and mine fades away, and that's ok too :-)

**Comment by borisfom** [comment]
No I definitely don't want to create new names. Most of the time what we need is a simple device to shadow outdated .cmake with the latest from Cmake (not right now, but I expect my fixes to be merged soon). My repo is not in a position now to replace .cmake files with FILE() references as I have new changes. But yours is :)

**Comment by borisfom** [comment]
That said, I withdraw this request!

**Pull request closed without merging**